src/charm.py

```diff
@@ -191,185 +190,101 @@ def unit_peer_data(self) -> Dict:
-            result = self.app_peer_data.get(key, None)
-
-        # TODO change upgrade to switch to secrets once minor version upgrades is done
-        if result:
-            return result
-
-        juju_version = JujuVersion.from_environ()
-        if juju_version.has_secrets:
-            return self._juju_secret_get_key(scope, key)
-
-    def _juju_secret_set(self, scope: str, key: str, value: str) -> Optional[str]:
-        """Helper function setting Juju secret."""
-        if scope == UNIT_SCOPE:
-            peer_data = self.unit_peer_data
-        else:
-            peer_data = self.app_peer_data
-        self._juju_secrets_get(scope)
-
-        key = SECRET_KEY_OVERRIDES.get(key, key)
-
-        secret = self.secrets[scope].get(SECRET_LABEL)
+        if value := self._peer_data(scope).get(key, None):
+            return value
 
-        # It's not the first secret for the scope, we can reuse the existing one
-        # that was fetched in the previous call
-        if secret:
-            secret_cache = self.secrets[scope][SECRET_CACHE_LABEL]
-
-            if secret_cache.get(key) == value:
-                logging.debug(f"Key {scope}:{key} has this value defined already")
-            else:
-                secret_cache[key] = value
-                try:
-                    secret.set_content(secret_cache)
-                except OSError as error:
-                    logging.error(
-                        f"Error in attempt to set {scope}:{key}. "
-                        f"Existing keys were: {list(secret_cache.keys())}. {error}"
-                    )
-                    return
-                logging.debug(f"Secret {scope}:{key} was {key} set")
-
-        # We need to create a brand-new secret for this scope
-        else:
-            scope_obj = self._scope_obj(scope)
+        if JujuVersion.from_environ().has_secrets:
+            secret_key = self._translate_field_to_secret_key(key)
+            label = generate_secret_label(self, scope)
+            for attempt in Retrying(stop=stop_after_attempt(3), wait=wait_fixed(1), reraise=True):
+                with attempt:
+                    secret = self.secrets.get(label)
 
-            secret = scope_obj.add_secret({key: value})
             if not secret:
-                raise RuntimeError(f"Couldn't set secret {scope}:{key}")
-
-            self.secrets[scope][SECRET_LABEL] = secret
-            self.secrets[scope][SECRET_CACHE_LABEL] = {key: value}
-            logging.debug(f"Secret {scope}:{key} published (as first). ID: {secret.id}")
-            peer_data.update({SECRET_INTERNAL_LABEL: secret.id})
-
-        # TODO change upgrade to switch to secrets once minor version upgrades is done
-        if key in peer_data:
-            del peer_data[key]
+                return
 
-        return self.secrets[scope][SECRET_LABEL].id
+            value = secret.get_content().get(secret_key)
+            if value != SECRET_DELETED_LABEL:
+                return value
```

Comment:
Why not do this in `self.secrets.get()`?

Reply:
I was thinking about this one too.

As long as hoping for unification of the `<charmnam>_secrets` module across charms, I'd like to leave this out.

The reason is: this is a Juju 3.1.6 > workaround.
Now we already have a bunch of charms that could drop Juju2. I'm wondering that they may be able 2 drop 3.1.5 at some point -- while other charms may still need to support it.

I'd rather not put a version-bound workaround in a lib, that has a chance for centralization.

It's a logical suggestion, but I'm wondering that --for now-- it may be better to keep it outside.
(If the moduel relamins local and OK to diverge, this could be refactored into the module)